core/src/main/java/org/apache/calcite/sql2rel/RelFieldTrimmer.java

```diff
@@ -297,11 +297,15 @@ protected final TrimResult dispatchTrimFields(
       assert newFieldCount > 0 : "rel has no fields after trim: " + rel;
     }
     if (newRel.equals(rel)) {
-      return result(rel, mapping);
+      return result(rel, mapping, rel);
     }
```

Comment:
Yes, we can do that.
In any case, it seems I made a mistake here, I shouldn't have used the 3rd parameters in this line, since both rels are the same.

Reply:
@danny0405 I can create a minor follow-up PR to deal with these details:
- Remove third parameter from RelFieldTrimmer:300
- Add early exit in `RelOptUtil.copyRelHints`:
```
public static RelNode copyRelHints(RelNode originalRel, RelNode newRel, boolean filterHints) {
  if (originalRel == newRel && !filterHints) {
    return originalRel;
  }
  ...
```

Reply:
Agree, thanks for your effort.

Reply:
[PR created](https://github.com/apache/calcite/pull/2831)